File: Algorithm_Checker.c
#include <cs50.h>
#include <stdio.h>

int adding(long card_number);
int multiply_add(int last_digit);
int card_length(long card_number);

int main(void)
{

    // get card number and display it
    long card_number = get_long("Enter Card Number : ");
    printf("Number : %li\n", card_number);
    int len = card_length(card_number);

    int checksum = adding(card_number);
    // display service provider
    if (checksum % 10 == 0)
    {
        // get starting digits
        do
        {
            card_number = card_number / 10;
        }
        while (card_number >= 100);

        // for AmEx
        if ((card_number == 34 || card_number == 37) && len == 15)
        {
            printf("AMEX\n");
        }
        else
        {
            printf("INVALID\n");
        }

        // for master card
        if (len == 16 && (card_number == 51 || card_number == 52 || card_number == 53 ||
                          card_number == 54 || card_number == 55))
        {
            printf("MASTERCARD\n");
        }

        // for visa, using if to resolve conflict with mastercard check
        if ((len == 13 || len == 16) &&
            (card_number == 40 || card_number == 41 || card_number == 42 || card_number == 43 ||
             card_number == 44 || card_number == 45 || card_number == 46 || card_number == 47 ||
             card_number == 48 || card_number == 49))
        {
            printf("VISA\n");
        }
    }
    else
    {
        printf("INVALID\n");
        return 0;
    }
}

// new function for checksum
int adding(long card_number)
{
    int sum_a = 0;
    bool isAlt = false;

    while (card_number > 0)
    {
        if (isAlt == true)
        {
            int last_digit = card_number % 10;
            int multi = multiply_add(last_digit);
            sum_a = sum_a + multi;
        }
        else
        {
            int last_digit = card_number % 10;
            sum_a = sum_a + last_digit;
        }

        isAlt = !isAlt;
        card_number = card_number / 10;
    }
    return sum_a;
}

// new function to multiply every even digit from the end by 2 and add them individually if they are
// double digits
int multiply_add(int last_digit)
{
    int sum_b = 0;
    int product = last_digit * 2;

    while (product > 0)
    {
        int last_multiplied_digit = product % 10;
        sum_b = sum_b + last_multiplied_digit;
        product = product / 10;
    }
    return sum_b;
}

// function to get the length of card number
int card_length(long card_number)
{
    int digits = 0;

    do
    {
        card_number = card_number / 10;
        digits++;
    }
    while (card_number != 0);
    return digits;
}

File: Algorithm_Checker.py
import re
while True :
    number = input("Number : ")
    if (bool(re.fullmatch(r"\d+", number)) == True):
        break
    print("INVALID")

card_number_length = len(str(number))
first_two_digits = int(number[:2])

digits_list = [int(d) for d in number][::-1]

sum_even = 0
sum_multiplied_odd = 0

for i in range(len(digits_list)):
    if i % 2 == 1:
        x = digits_list[i] * 2
        sum_multiplied_odd += ((x % 10) + (x // 10))
    else:
        sum_even += digits_list[i]


checksum = sum_even + sum_multiplied_odd
last_digit_checksum = checksum % 10

if checksum % 10 == 0:

    if((first_two_digits == 34 or first_two_digits == 37) and card_number_length == 15):
        print("AMEX")
    elif((first_two_digits == 51 or first_two_digits == 52 or first_two_digits == 53 or first_two_digits == 54 or first_two_digits == 55) and card_number_length == 16):
        print("MASTERCARD")
    elif(int(number[:1]) == 4 and (card_number_length == 13 or card_number_length ==16)):
        print("VISA")
    else:
        print("INVALID")
else:
    print("INVALID")
